File: src/languages/nama/nama.ts
/**
 * Nama / Khoekhoe (naq) phonemizer — a greedy scan over the Khoekhoegowab click orthography, canonical
 * IPA. This file owns the click composition (place × efflux is a rule, not a 20-row table — clickIPA
 * below), the ⟨kh⟩ digraph, doubled-vowel length, and the word-final gender-⟨-b⟩ devoicing. The non-click
 * letter table and the encyclopedic record (the click system, the spec-verification caveat) live in
 * nama.jsonc.
 */
import type { Phonemizer } from "../../registry.ts";
import { assembleClauses } from "../../core/clauses.ts";
import { LATIN_RUN, makeNativiser } from "../../core/hostWord.ts";
import { loadManifest } from "../../core/loadManifest.ts";
import { numberToWords, readDigits } from "./numbers.ts";
import { latinPhone } from "../../core/latinPhones.ts";
import { normalizeNama } from "./normalize.ts";

interface NamaDef {
    clicks: readonly string[];
    plainVowels: readonly string[];
    letters: Record<string, string>;
}
const DEF = loadManifest<NamaDef>(import.meta.url, "nama.jsonc");
const CLICK = new Set(DEF.clicks); // dental, lateral, palatal, alveolar
/** A click letter + its accompaniment (the following g/kh/h/n, longest-first) → the IPA click cluster. */
function clickIPA(click: string, accomp: string): string {
    switch (accomp) {
        case "g": return "ᵏ" + click; // tenuis (voiceless unaspirated)
        case "kh": return "ᵏ" + click + "ʰ"; // aspirated
        case "h": return "ᵑ̊" + click + "ʰ"; // aspirated (voiceless) nasal
        case "n": return "ᵑ" + click; // voiced nasal
        default: return "ᵑ̊" + click + "ˀ"; // BARE click → the glottalised nasal click
    }
}
// Non-click letters (nama.jsonc). ⟨kh⟩→[kʰ] handled as a digraph; ⟨g⟩ NOT after a click → [x].
const LETTER = DEF.letters;
const PLAIN_VOWEL = new Set(DEF.plainVowels);

/** One Nama word → canonical IPA. */
export function phonemizeWord(word: string): string {
    const chars = [...word.normalize("NFC")];
    // Lowercase PER-INDEX (not a separate array) so a char that lowercases to a different code-point count (İ→i̇, ß→ss)
    // can never desync the index from `chars`.
    const lc = (idx: number): string => (chars[idx] ?? "").toLowerCase();
    const segs: string[] = [];
    for (let i = 0; i < chars.length; i++) {
        const c = chars[i]!;
        if (CLICK.has(c)) {
            // Accompaniment: longest-first — ⟨kh⟩ (2 letters) before ⟨g h n⟩. Case-insensitive (ǀKh, ǀG in citations).
            const next2 = lc(i + 1) + lc(i + 2);
            const next1 = lc(i + 1);
            if (next2 === "kh") { segs.push(clickIPA(c, "kh")); i += 2; }
            else if (next1 === "g" || next1 === "h" || next1 === "n") { segs.push(clickIPA(c, next1)); i += 1; }
            else segs.push(clickIPA(c, ""));
            continue;
        }
        const cur = lc(i);
        if (cur === "k" && lc(i + 1) === "h") { segs.push("kʰ"); i++; continue; } // ⟨kh⟩ digraph
        // A DOUBLED identical vowel → a long vowel [Vː] (the standard Khoekhoegowab length convention, aa/ee…).
        if (PLAIN_VOWEL.has(cur) && lc(i + 1) === cur) { segs.push(LETTER[cur]! + "ː"); i++; continue; }
        // WORD-FINAL gender suffix ⟨-b⟩ → [p] (devoicing): ǀgomab→ǀómàp.
        if (cur === "b" && i === chars.length - 1) { segs.push("p"); continue; }
        // ⚠ A letter with no rule here still denotes a sound; dropping it deletes what the writer typed.
        // Reached only when every rule above has declined, so the language's own reading always wins.
        const ph = LETTER[cur] ?? latinPhone(cur, { initial: i === 0, includeH: true });
        if (ph !== undefined) segs.push(ph);
        // tone diacritics on vowels, ʼ, etc.: dropped (tone not emitted; it folds)
    }
    return segs.join("").normalize("NFC");
}

// Nama Latin + the click letters ǀ ǁ ǂ ǃ. Word / number / punctuation.
/**
 * This language's OWN inventory. ⚠ TWO DIFFERENT QUESTIONS, KEPT APART: the TOKEN class below decides where
 * the SCRIPT boundary falls (routing), while this one decides whether the g2p has rules for these letters. A
 * token this class REJECTS carries a letter the language does not use — i.e. a foreign name, which `nat`
 * then folds to a base the g2p does have a rule for. See core/hostWord.ts.
 */
/**
 * ⚠ THE MACRON AND CIRCUMFLEX VOWELS ARE IN THE CLASS, and leaving them out cost this language BOTH of its
 * diacritic contrasts (#1140). `nama.jsonc`'s own `letters` table declares them and says what they are:
 * macron = LONG (ā → aː), circumflex = NASALIZED, annotated there as phonemic (ǂgâ, ǀî). But a token outside
 * this class is FOLDED, and `core/hostWord.ts` strips combining marks — so ⟨ā⟩ and ⟨â⟩ both arrived as ⟨a⟩
 * and both contrasts were erased before the g2p ran: `ǃkhās` read *ᵏǃʰas* not *ᵏǃʰaːs*, `ǂgâ` read *ᵏǂa* not
 * *ᵏǂã*. Nothing looked broken, because a folded letter still produces a sound.
 *
 * ⚠ naq HAS NO GOLDEN AND NO CORPUS ARTIFACT in this repo, so no differential can witness this and the tests
 * are the whole instrument. Both cases are listed because the nativiser flags here are "u", not "iu".
 */
const NATIVE_CLASS = "[a-zA-ZāēīōūâêîôûĀĒĪŌŪÂÊÎÔÛǀǁǂǃ]";
const nat = makeNativiser(NATIVE_CLASS, "u");

// ⚠ ALL OF LATIN, not just this language's own letters — the narrow class ended the token at an
// out-of-inventory diacritic, so that letter became an unclaimed gap read as an English LETTER NAME and the
// rest of the word started over: `São Paulo` fragmented into three pieces, none of them right. Invisible to
// every gate: no digit or raw mark survives and nothing VANISHES.
const TOKEN = new RegExp(`(${LATIN_RUN})|(\\d+)|([.?!,;:…])`, "gu");

class NamaPhonemizer implements Phonemizer {
    text(input: string): string {
        return assembleClauses(normalizeNama(input.normalize("NFC")), TOKEN, (m, sink) => {
            if (m[1]) sink.emit(phonemizeWord(nat(m[1])));
            else if (m[2]) {
                // Cardinals 1 … 10¹²−1 compose natively; 0 emits the flagged Afrikaans stopgap `nul` and anything
                // above the ceiling reads digit-by-digit. Never silently dropped. See numbers.ts.
                const words = m[2].length <= 12 ? numberToWords(Number(m[2]), m[2]) : readDigits(m[2]);
                for (const wd of words.split(" ")) sink.emit(phonemizeWord(wd));
            }
            else if (m[3]) sink.pause(m[3] === "." || m[3] === "!" || m[3] === "?" ? m[3] : ",");
        });
    }
}

/** Build the Nama (Khoekhoe) phonemizer (click-aware Khoekhoegowab scan). */
export function createNama(): Phonemizer {
    return new NamaPhonemizer();
}

File: csharp/Vernacula.Phonemizer/Languages/Nama/Nama.cs
/**
 * Nama / Khoekhoe (naq) phonemizer — a greedy scan over the Khoekhoegowab click orthography, canonical IPA.
 * Owns the click composition (place × efflux is a rule, not a table), the ⟨kh⟩ digraph, doubled-vowel
 * length, and the word-final gender-⟨-b⟩ devoicing.
 * Ported from src/languages/nama/nama.ts — see that file for the corpus evidence.
 */
using System.Text;
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Nama;

public sealed class NamaPhonemizer : ILanguage
{
    private static readonly IReadOnlyDictionary<string, string> LETTER = Manifest.DEF.Letters;

    /** A click letter + its accompaniment (the following g/kh/h/n, longest-first) → the IPA click cluster. */
    private static string ClickIPA(string click, string accomp) => accomp switch
    {
        "g" => "ᵏ" + click, // tenuis (voiceless unaspirated)
        "kh" => "ᵏ" + click + "ʰ", // aspirated
        "h" => "ᵑ̊" + click + "ʰ", // aspirated (voiceless) nasal
        "n" => "ᵑ" + click, // voiced nasal
        _ => "ᵑ̊" + click + "ˀ", // BARE click → the glottalised nasal click
    };

    /** One Nama word → canonical IPA. */
    public static string PhonemizeWord(string word)
    {
        var chars = Js.CodePoints(Js.Normalize(word, NormalizationForm.FormC));
        // Lowercase PER-INDEX (not a separate list) so a char that lowercases to a different code-point
        // count (İ→i̇, ß→ss) can never desync the index from `chars`.
        static string Lc(List<string> cs, int idx) => idx < cs.Count ? Js.ToLowerCase(cs[idx]) : "";
        var segs = new List<string>();
        for (var i = 0; i < chars.Count; i++)
        {
            var c = chars[i];
            if (Manifest.CLICK.Contains(c))
            {
                // Accompaniment: longest-first — ⟨kh⟩ (2 letters) before ⟨g h n⟩. Case-insensitive (ǀKh, ǀG in citations).
                var next2 = Lc(chars, i + 1) + Lc(chars, i + 2);
                var next1 = Lc(chars, i + 1);
                if (next2 == "kh") { segs.Add(ClickIPA(c, "kh")); i += 2; }
                else if (next1 is "g" or "h" or "n") { segs.Add(ClickIPA(c, next1)); i += 1; }
                else segs.Add(ClickIPA(c, ""));
                continue;
            }
            var cur = Lc(chars, i);
            if (cur == "k" && Lc(chars, i + 1) == "h") { segs.Add("kʰ"); i++; continue; } // ⟨kh⟩ digraph
            // A DOUBLED identical vowel → a long vowel [Vː] (the standard Khoekhoegowab length convention).
            if (Manifest.PLAIN_VOWEL.Contains(cur) && Lc(chars, i + 1) == cur) { segs.Add(LETTER[cur] + "ː"); i++; continue; }
            // WORD-FINAL gender suffix ⟨-b⟩ → [p] (devoicing).
            if (cur == "b" && i == chars.Count - 1) { segs.Add("p"); continue; }
            // A letter with no rule here still denotes a sound; dropping it deletes what the writer typed.
            var ph = LETTER.TryGetValue(cur, out var l) ? l : LatinPhones.LatinPhone(cur, new PhoneOpts { Initial = i == 0, IncludeH = true });
            if (ph != null) segs.Add(ph);
            // tone diacritics on vowels, ʼ, etc.: dropped (tone not emitted; it folds)
        }
        return Js.Normalize(string.Join("", segs), NormalizationForm.FormC);
    }

    /**
     * This language's OWN inventory — the macron and circumflex vowels ARE in the class (leaving them out
     * folded both diacritic contrasts to their bare bases before the g2p ran).
     */
    private static readonly Func<string, string> NAT =
        HostWord.MakeNativiser("[a-zA-ZāēīōūâêîôûĀĒĪŌŪÂÊÎÔÛǀǁǂǃ]", "u");

    // Nama Latin + the click letters ǀ ǁ ǂ ǃ. Word / number / punctuation.
    // ⚠ ALL OF LATIN in the word arm, not just this language's own letters.
    private static readonly JsRe TOKEN =
        JsRegex.Compile($"({HostWord.LATIN_RUN})|(\\d+)|([.?!,;:…])", "gu");

    public string Text(string input)
    {
        return Clauses.AssembleClauses(Normalize.NormalizeNama(Rewriter.Renormalize(input, NormalizationForm.FormC)), TOKEN, (m, sink) =>
        {
            if (m.Groups[1].Success) sink.Emit(PhonemizeWord(NAT(m.Groups[1].Value)));
            else if (m.Groups[2].Success)
            {
                // Cardinals 1 … 10¹²−1 compose natively; 0 emits the flagged Afrikaans stopgap `nul` and
                // anything above the ceiling reads digit-by-digit. Never silently dropped.
                var v = m.Groups[2].Value;
                var words = v.Length <= 12 ? Numbers.NumberToWords(Js.Number(v), v) : Numbers.ReadDigits(v);
                foreach (var wd in words.Split(' ')) sink.Emit(PhonemizeWord(wd));
            }
            else if (m.Groups[3].Success)
            {
                var p = m.Groups[3].Value;
                sink.Pause(p is "." or "!" or "?" ? p : ",");
            }
        });
    }

    /** Build the Nama (Khoekhoe) phonemizer (click-aware Khoekhoegowab scan). */
    public static ILanguage CreateNama() => new NamaPhonemizer();

    internal static void RegisterSelf() => Registry.Register("nama", CreateNama);
}
